include/souper/Infer/InstSynthesis.h

```diff
@@ -0,0 +1,300 @@
+  method in Souper.
+*/
+
+namespace souper {
+
+/// A location variable identifies an input variable, the output variable, or
+/// input/output of a library component. The encoding has the following
+/// semantics: If the first integer is 0, it's an input var or a constant
+/// where the second integer describes its number. If the second integer is 0,
+/// it's an output var and the first integer describes the number of a
+/// component. The output variable is encoded by second integer being 0 and
+/// the first integer being equal to the number of library components plus one.
+/// Otherwise, it's a component input where the first integer describes the
+/// number of a component and the second integer describes it's input number.
+/// Note that constant components (no inputs, one output) are not treated as
+/// ordinary library components, but rather as inputs.
+/// For example:
+/// 0_1, 0_2, 0_3 describe the input variables 1, 2, and 3.
+/// 1_0, 2_0, 3_0 describe the outputs of components 1, 2, and 3.
+/// 1_1, 2_2, 3_1 describe the first input of component one, the second input of
+/// component two, and the first input of component three.
+/// 4_0 describes the final output (assuming there are only three components)
+typedef std::pair<unsigned, unsigned> LocVar;
+
+/// A mapping from a location variable to it's concrete value (Inst::Const).
+/// The values of location variables decide the interconnections between the
+/// various components. To describe a program, we have to determine which
+/// component goes on which location (line-number), and from which location
+/// (line-number of program input) does it get its input arguments. E.g.,
+/// the location mappings
+/// 0_1 -> 0
+/// 1_1 -> 0
+/// 1_0 -> 1
+/// 2_0 -> 1
+/// describe a program with one input (0_1) and one component with one
+/// input (1_1) located at line-number 0, producing output (1_0) at line-number 1
+/// which is the output of the program (2_0).
+/// Initially, the values of location variables are symbolic and it is the job
+/// of the SMT solver to figure out a valid wiring representing a valid program
+typedef std::pair<LocVar, Inst *> LocInst;
+
+/// A component is a fixed-width instruction kind
+struct Component {
+  Inst::Kind Kind;
+  unsigned Width;
+  std::vector<unsigned> OpWidths;
+};
+
+/// Unsupported components kinds
+static const std::set<Inst::Kind> UnsupportedCompKinds = {
+  Inst::Var, Inst::UntypedConst,
+  // Not implemented
+  Inst::Phi,
+  // Disabled due to scalability challenges
+  Inst::AddNSW, Inst::AddNUW, Inst::AddNW,
+  Inst::SubNSW, Inst::SubNUW, Inst::SubNW,
+  Inst::MulNSW, Inst::MulNUW, Inst::MulNW,
+  Inst::ShlNSW, Inst::ShlNUW, Inst::ShlNW,
+};
+
+/// Supported component library.
+/// The width of most of the instructions is unknown in advance, therefore, we
+/// indicate this by setting the Width to 0. During initialization, the width
+/// will be set to the DefaultWidth (maximum width of the input vars).
```

Comment:
At the moment, definitely too expensive. Let's consider this once the performance of synthesis significantly improves.